app/src/androidTest/java/com/lightricks/feedexercise/data/FeedRepositoryTest.kt

```diff
@@ -1,13 +1,163 @@
+import io.reactivex.Single
+import org.junit.After
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
 import org.junit.runner.RunWith
-import java.util.concurrent.CountDownLatch
-import java.util.concurrent.TimeUnit
+import java.io.IOException
 
 @RunWith(AndroidJUnit4::class)
 class FeedRepositoryTest {
-   //todo: add the tests here
+
+    private lateinit var userDao: FeedItemDao
+    private lateinit var db: FeedDatabase
+    private lateinit var appContext: Context
+    private lateinit var jsonContent: String
+    private lateinit var serviceResponse: TemplatesMetadata
+
+    @Before
+    fun createDb() {
+        appContext = ApplicationProvider.getApplicationContext<Context>()
+        db = Room.inMemoryDatabaseBuilder(
+            appContext, FeedDatabase::class.java).build()
+        userDao = db.feedItemDao()
+    }
+
+    @Before
+    fun getJsonStr(){
+        jsonContent = appContext.assets
+            .open("get_feed_response.json")
+            .bufferedReader()
+            .use{ it.readText() }
+        val moshi = Moshi.Builder()
+            .add(KotlinJsonAdapterFactory())
+            .build()
+        val adapter = moshi.adapter<TemplatesMetadata>(TemplatesMetadata::class.java)
+        serviceResponse = adapter.fromJson(jsonContent)!!
+    }
+
+    @After
+    @Throws(IOException::class)
+    fun closeDb() {
+        db.close()
+    }
+
+    @get:Rule
+    var instantExecutorRule = InstantTaskExecutorRule()
+
+    @Test
+    fun refreshShouldFailOnBadNetworkAddress(){
+        FeedApiService.BASE_URL="http://badaddress.baddomain"
+        val service =
+            FeedApiService.FeedApi.service
+        val repository = FeedRepository(db, service)
+        val viewModel = FeedViewModel(repository)
+        var errorString=""
+
+        val errorObserver = object : Observer<Event<String>> {
+            override fun onChanged(t: Event<String>?) {
+                if (t != null)
+                    errorString = t.peekContent()
+                }
+        }
+        viewModel.getNetworkErrorEvent().observeForever (errorObserver)
+
+        waitForEndOfLoading(viewModel)
+        viewModel.getNetworkErrorEvent().removeObserver(errorObserver)
+        assert(errorString != "")
+
+    }
+
+    @Test
+    fun refreshShouldSaveItemsToDB(){
+        val service = object: FeedApiService{
+            override fun getFeed(): Single<TemplatesMetadata> {
+                return Single.just(serviceResponse)
+            }
+        }
+        val itemListSize=Integer(serviceResponse.templatesMetadata.size)
+        val repository = FeedRepository(db, service)
+        val observer = repository.refresh().test()
+
+        observer.awaitTerminalEvent()
+        observer
+            .assertNoErrors()
+            .assertComplete()
+        userDao
+            .getCount().test()
+            .assertValues(itemListSize)
+    }
+
+    @Test
+    fun feedItemsShouldContainDBContent() {
+        val mockID = "ID"
```

Comment:
Let's make the test simpler by using the following algorithm:
1. Read the items using `repository.feedItems.blockingObserve()` and check that they are empty.
2. Insert some items into `FeedItemDao` directly (not via the repository).
3. Read the items using `repository.feedItems.blockingObserve()` and verify that the items match the inserted items